allowlist.go

```diff
@@ -0,0 +1,190 @@
+package rcmgr
+
+import (
+	"bytes"
+	"errors"
+	"net"
+
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/multiformats/go-multiaddr"
+	manet "github.com/multiformats/go-multiaddr/net"
+)
+
+type allowlist struct {
+	// a simple structure of lists of networks. There is probably a faster way
+	// to check if an IP address is in this network than iterating over this
+	// list, but this is good enough for small numbers of networks (<1_000).
+	// Analyze the benchmark before trying to optimize this.
+
+	// Any peer with these IPs are allowed
+	allowedNetworks []*net.IPNet
+
+	// Only the specified peers can use these IPs
+	allowedPeerByNetwork map[peer.ID][]*net.IPNet
```

Comment:
Have you considered unifying these into a single map? `map[net.IPNet]struct{ allowsAll bool; map[peer.ID]struct{} }`
`allowsAll` would be set if a multiaddr without /p2p is added.

Reply:
Can I use net.IPNet as a key? The struct holds two byteslices, so I don't think I can.

Reply:
Just checked, you can't. That's annoying. You _could_ use `net.IPNet.String()`, but not sure if that's nicer.

Reply:
I don’t think so because:

1. allow checking is harder. you have to convert between the string and bytes.
2. You need to allocate the string on insertion. 
3. you need to allocate at least once when doing an allowlist check. 
4. (Depending on check implementation) you lose the quick lookup by peerid. Which may be a pretty common use case.